middleware/csrf.go

```diff
@@ -0,0 +1,234 @@
+package middleware
+
+import (
+	"crypto/rand"
+	"crypto/subtle"
+	"encoding/base64"
+	"errors"
+	"github.com/gobuffalo/buffalo"
+	"net/http"
+	"net/url"
+)
+
+const (
+	// CSRF token length in bytes.
+	csrfTokenLength int    = 32
+	csrfTokenKey    string = "authenticity_token"
+)
+
+var (
+	// The name value used in form fields.
+	fieldName = csrfTokenKey
+
+	// The HTTP request header to inspect
+	headerName = "X-CSRF-Token"
+
+	// Idempotent (safe) methods as defined by RFC7231 section 4.2.2.
+	safeMethods = []string{"GET", "HEAD", "OPTIONS", "TRACE"}
+)
+
+var (
+	// ErrNoReferer is returned when a HTTPS request provides an empty Referer
+	// header.
+	ErrNoReferer = errors.New("referer not supplied")
+	// ErrBadReferer is returned when the scheme & host in the URL do not match
+	// the supplied Referer header.
+	ErrBadReferer = errors.New("referer invalid")
+	// ErrNoToken is returned if no CSRF token is supplied in the request.
+	ErrNoToken = errors.New("CSRF token not found in request")
+	// ErrBadToken is returned if the CSRF token in the request does not match
+	// the token in the session, or is otherwise malformed.
+	ErrBadToken = errors.New("CSRF token invalid")
+)
+
+// EnableCSRF enable CSRF protection on routes using this middleware.
+// This middleware is adapted from gorilla/csrf
+func EnableCSRF() buffalo.MiddlewareFunc {
```

Comment:
Rename to `CSRF` or `CSRFProtection` something like that.

Also, there's no need to return a `buffalo.MiddlewareFunc` here, just make the function itself implement it.

```go
func CSRF(next buffalo.Handler) buffalo.Handler {
...
}
```